### ISSUE TITLE
GUIEffect: Issue with transitive @UI inheritance

### ISSUE THREAD

**Issue by lazaroclapp** [typetools/checker-framework#2049]
We have observed some strange behavior for the GUIEffect checker when dealing with long (>2 classes deep) inheritance hierarchies.

Consider the following example code:

```
class MyTestCase {

    @PolyUIType
    static class PolyUIBaseClass {
        @PolyUIEffect
        public void foo() {};
    }

    static class TransitiveUISubclass extends @UI PolyUIBaseClass { }

    static class UISubclass extends TransitiveUISubclass {
        @UIEffect
        public void bar() {}
        @Override
        public void foo() {
            bar();
        }
    }

}
```

Unless I am mistaken, I would expect the `guieffect` checker on this code to compile without error (which is what happens if `UISubclass` is made to extend `@UI PolyUIBaseClass` directly, except for a `effects.redundant.uitype` warning), or else to give an override error, where it complains that `foo` should be `@UIEffect`, as it overrides `@UIEffect TransitiveUISubclass.foo()`.

However, instead, this code fails with `error: (call.invalid.ui)` on the callsite for `bar();` inside `foo`, showing that it considers the declaration of `UISubclass.foo` to be both valid and `@SafeEffect`.

Is this expected behavior or an error in the checker?

cc: @msridhar @csgordon 

p.s. This is not a high urgency issue for us right now, but it seems very unintuitive and did mean that in the past we weren't checking some things that we thought we were.

**Comment by csgordon** [comment]
Thanks for the report.  That is indeed incorrect.  I'll try to fix this sooner than later.

**Comment by csgordon** [comment]
Digging up this bug again after a long time... this behavior actually looks correct (or, at least it's consistent with my intent when I wrote the code).  This gets flagged because the default effect for ```UISubclass.foo()``` is ```@SafeEffect```.  Explicitly marking the method ```@UIEffect``` (or marking the class ```@UIType```) makes this code pass.

If I change the subclass to inherit directly from ```@UI PolyUIBaseClass```, contrary to what's suggested in the report, this is still flagged.  Only if this ```foo()``` implementation was directly inside an anonymous inner class inheriting from ```@UI PolyUIBaseClass``` would the current design infer the intended effect of the method to be ```@UIEffect```.

Here's a variant where it does lead to an unintended error message:
```java

import org.checkerframework.checker.guieffect.qual.*;


class TransitivityBug2049 {

    @PolyUIType
    static class PolyUIBaseClass {
        @PolyUIEffect
        public void foo() {};
    }

    static class TransitiveUISubclass extends @UI PolyUIBaseClass { }

    static class UISubclass extends @UI PolyUIBaseClass {
        @UIEffect
        public void bar() {}
        @Override
        @UIEffect
        public void foo() {
            bar(); // Should be okay; Issue 2049 reported that this was incorrectly flagged
        }
    }

}
```
This leads to:
```
checker/tests/guieffect/TransitivityBug2049.java:19: error: [override.effect.invalid.nonui] A method override may only be @UI if it overrides an @UI method (overriding non-UI instantiation of supertype).
        public void foo() {
                    ^
    foo() in TransitivityBug2049.UISubclass
    cannot override
    foo() in @AlwaysSafe PolyUIBaseClass
```
The problem is that the ```@UI``` annotation in ```UISubclass```'s extends clause gets lost in checking the override.  This *is* a bug.
